// intern/lib/common/time.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@theintern/common"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var common_1 = require("@theintern/common");
    var _setTimeout = common_1.global.setTimeout;
    var _clearTimeout = common_1.global.clearTimeout;
    var _setInterval = common_1.global.setInterval;
    var _clearInterval = common_1.global.clearInterval;
    var now = Date.now;
    function createProxy(func, context) {
        return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return func.call.apply(func, [context].concat(args));
        };
    }
    var proxySetTimeout = createProxy(_setTimeout, common_1.global);
    exports.setTimeout = proxySetTimeout;
    var proxySetInterval = createProxy(_setInterval, common_1.global);
    exports.setInterval = proxySetInterval;
    var proxyClearTimeout = createProxy(_clearTimeout, common_1.global);
    exports.clearTimeout = proxyClearTimeout;
    var proxyClearInterval = createProxy(_clearInterval, common_1.global);
    exports.clearInterval = proxyClearInterval;
    var proxyNow = createProxy(now, Date);
    exports.now = proxyNow;
});
