// grunt-githash/Gruntfile.js
/*
 * grunt-githash
 * https://github.com/jantimon/grunt-githash
 *
 * Copyright (c) 2015 Jan Nicklas
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {
  // load all npm grunt tasks
  require('load-grunt-tasks')(grunt);

  // Project configuration.
  grunt.initConfig({
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>'
      ],
      options: {
        jshintrc: '.jshintrc'
      }
    },

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp']
    },

    // Configuration to be run (and then tested).
    githash: {
      main: {
        options: {
          dir: '.',
          fail: true
        }
      }
    },

    'template': {
      'githash-test': {
        'options': {
          'data': {
            'output': '<%= githash.main.hash %>'
          }
        },
        'files': {
          'tmp/result.txt': ['test/fixtures/test.tpl']
        }
      }
    },

    // Unit tests.
    nodeunit: {
      tests: ['test/*_test.js']
    }

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'githash', 'template', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'test']);

};
